File: test/language/choice_types/java/choice_types/IntChoiceTest.java
package choice_types;

import static org.junit.Assert.*;

import java.io.File;
import java.io.IOException;

import javax.imageio.stream.FileImageOutputStream;

import org.junit.Test;

import choice_types.int_choice.IntChoice;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.BitStreamWriter;
import zserio.runtime.io.FileBitStreamReader;
import zserio.runtime.io.FileBitStreamWriter;

public class IntChoiceTest
{
    @Test
    public void containerConstructor()
    {
        final int tag = VARIANT_A_SELECTOR;
        final IntChoice intChoice = new IntChoice(tag);
        assertEquals(tag, intChoice.getTag());
    }

    @Test
    public void fileConstructor() throws IOException, ZserioError
    {
        final int tag = VARIANT_A_SELECTOR;
        final File file = new File("test.bin");
        final int value = 99;
        writeIntChoiceToFile(file, tag, value);
        final IntChoice intChoice = new IntChoice(file, tag);
        assertEquals(tag, intChoice.getTag());
        assertEquals((byte)value, intChoice.getA());
    }

    @Test
    public void bitStreamReaderConstructor() throws IOException, ZserioError
    {
        final int tag = VARIANT_B_SELECTOR1;
        final File file = new File("test.bin");
        final int value = 234;
        writeIntChoiceToFile(file, tag, value);
        final BitStreamReader stream = new FileBitStreamReader(file);
        final IntChoice intChoice = new IntChoice(stream, tag);
        stream.close();
        assertEquals(tag, intChoice.getTag());
        assertEquals((short)value, intChoice.getB());
    }

    @Test
    public void bitSizeOf()
    {
        final IntChoice intChoiceA = new IntChoice(VARIANT_A_SELECTOR);
        final byte byteValueA = 99;
        intChoiceA.setA(byteValueA);
        assertEquals(8, intChoiceA.bitSizeOf());

        final IntChoice intChoiceB = new IntChoice(VARIANT_B_SELECTOR2);
        final short shortValueB = 234;
        intChoiceB.setB(shortValueB);
        assertEquals(16, intChoiceB.bitSizeOf());

        final IntChoice intChoiceEmpty = new IntChoice(EMPTY_SELECTOR1);
        assertEquals(0, intChoiceEmpty.bitSizeOf());

        final IntChoice intChoiceC = new IntChoice(VARIANT_C_SELECTOR);
        final int intValueC = 65535;
        intChoiceC.setC(intValueC);
        assertEquals(32, intChoiceC.bitSizeOf());
    }

    @Test
    public void getTag()
    {
        final int tag = EMPTY_SELECTOR2;
        final IntChoice intChoice = new IntChoice(tag);
        assertEquals(tag, intChoice.getTag());
    }

    @Test
    public void getSetA()
    {
        final IntChoice intChoice = new IntChoice(VARIANT_A_SELECTOR);
        final byte value = 99;
        intChoice.setA(value);
        assertEquals(value, intChoice.getA());
    }

    @Test
    public void getSetB()
    {
        final IntChoice intChoice = new IntChoice(VARIANT_B_SELECTOR3);
        final short value = 234;
        intChoice.setB(value);
        assertEquals(value, intChoice.getB());
    }

    @Test
    public void getSetC()
    {
        final IntChoice intChoice = new IntChoice(VARIANT_C_SELECTOR);
        final int value = 65535;
        intChoice.setC(value);
        assertEquals(value, intChoice.getC());
    }

    @Test
    public void equals()
    {
        final IntChoice intChoice1 = new IntChoice(VARIANT_A_SELECTOR);
        final IntChoice intChoice2 = new IntChoice(VARIANT_A_SELECTOR);
        assertTrue(intChoice1.equals(intChoice2));

        final byte value = 99;
        intChoice1.setA(value);
        assertFalse(intChoice1.equals(intChoice2));

        intChoice2.setA(value);
        assertTrue(intChoice1.equals(intChoice2));

        final byte diffValue = value + 1;
        intChoice2.setA(diffValue);
        assertFalse(intChoice1.equals(intChoice2));
    }

    @Test
    public void hashCodeMethod()
    {
        final IntChoice intChoice1 = new IntChoice(VARIANT_A_SELECTOR);
        final IntChoice intChoice2 = new IntChoice(VARIANT_A_SELECTOR);
        assertEquals(intChoice1.hashCode(), intChoice2.hashCode());

        final byte value = 99;
        intChoice1.setA(value);
        assertTrue(intChoice1.hashCode() != intChoice2.hashCode());

        intChoice2.setA(value);
        assertEquals(intChoice1.hashCode(), intChoice2.hashCode());

        final byte diffValue = value + 1;
        intChoice2.setA(diffValue);
        assertTrue(intChoice1.hashCode() != intChoice2.hashCode());
    }

    @Test
    public void initializeOffsets()
    {
        final int bitPosition = 1;
        final IntChoice intChoiceA = new IntChoice(VARIANT_A_SELECTOR);
        assertEquals(9, intChoiceA.initializeOffsets(bitPosition));

        final IntChoice intChoiceB = new IntChoice(VARIANT_B_SELECTOR2);
        assertEquals(17, intChoiceB.initializeOffsets(bitPosition));

        final IntChoice intChoiceEmpty = new IntChoice(EMPTY_SELECTOR1);
        assertEquals(1, intChoiceEmpty.initializeOffsets(bitPosition));

        final IntChoice intChoiceC = new IntChoice(VARIANT_C_SELECTOR);
        assertEquals(33, intChoiceC.initializeOffsets(bitPosition));
    }

    @Test
    public void fileWrite() throws IOException, ZserioError
    {
        final IntChoice intChoiceA = new IntChoice(VARIANT_A_SELECTOR);
        final byte byteValueA = 99;
        intChoiceA.setA(byteValueA);
        final File file = new File("test.bin");
        intChoiceA.write(file);
        final IntChoice readIntChoiceA = new IntChoice(file, VARIANT_A_SELECTOR);
        assertEquals(byteValueA, readIntChoiceA.getA());

        final IntChoice intChoiceB = new IntChoice(VARIANT_B_SELECTOR1);
        final short shortValueB = 234;
        intChoiceB.setB(shortValueB);
        intChoiceB.write(file);
        final IntChoice readIntChoiceB = new IntChoice(file, VARIANT_B_SELECTOR1);
        assertEquals(shortValueB, readIntChoiceB.getB());

        final IntChoice intChoiceC = new IntChoice(VARIANT_C_SELECTOR);
        final int intValueC = 65535;
        intChoiceC.setC(intValueC);
        intChoiceC.write(file);
        final IntChoice readIntChoiceC = new IntChoice(file, VARIANT_C_SELECTOR);
        assertEquals(intValueC, readIntChoiceC.getC());
    }

    @Test
    public void bitStreamWriterWrite() throws IOException, ZserioError
    {
        final IntChoice intChoiceA = new IntChoice(VARIANT_A_SELECTOR);
        final byte byteValueA = 99;
        intChoiceA.setA(byteValueA);
        final File file = new File("test.bin");
        BitStreamWriter writer = new FileBitStreamWriter(file);
        intChoiceA.write(writer);
        writer.close();
        final IntChoice readIntChoiceA = new IntChoice(file, VARIANT_A_SELECTOR);
        assertEquals(byteValueA, readIntChoiceA.getA());

        final IntChoice intChoiceB = new IntChoice(VARIANT_B_SELECTOR1);
        final short shortValueB = 234;
        intChoiceB.setB(shortValueB);
        writer = new FileBitStreamWriter(file);
        intChoiceB.write(writer);
        writer.close();
        final IntChoice readIntChoiceB = new IntChoice(file, VARIANT_B_SELECTOR1);
        assertEquals(shortValueB, readIntChoiceB.getB());

        final IntChoice intChoiceC = new IntChoice(VARIANT_C_SELECTOR);
        final int intValueC = 65535;
        intChoiceC.setC(intValueC);
        writer = new FileBitStreamWriter(file);
        intChoiceC.write(writer);
        writer.close();
        final IntChoice readIntChoiceC = new IntChoice(file, VARIANT_C_SELECTOR);
        assertEquals(intValueC, readIntChoiceC.getC());
    }

    private void writeIntChoiceToFile(File file, int tag, int value) throws IOException
    {
        final FileImageOutputStream stream = new FileImageOutputStream(file);

        switch (tag)
        {
        case 1:
            stream.writeByte(value);
            break;

        case 2:
        case 3:
        case 4:
            stream.writeShort(value);
            break;

        case 5:
        case 6:
            break;

        default:
            stream.writeInt(value);
        }

        stream.close();
    }

    private static int VARIANT_A_SELECTOR = 1;
    private static int VARIANT_B_SELECTOR1 = 2;
    private static int VARIANT_B_SELECTOR2 = 3;
    private static int VARIANT_B_SELECTOR3 = 4;
    private static int EMPTY_SELECTOR1 = 5;
    private static int EMPTY_SELECTOR2 = 6;
    private static int VARIANT_C_SELECTOR = 7;
}

File: test/language/choice_types/cpp/IntChoiceTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "choice_types/int_choice/IntChoice.h"

namespace choice_types
{
namespace int_choice
{

class IntChoiceTest : public ::testing::Test
{
protected:
    void writeIntChoiceToByteArray(zserio::BitStreamWriter& writer, uint16_t tag, int32_t value)
    {
        switch (tag)
        {
        case 1:
            writer.writeSignedBits(value, 8);
            break;

        case 2:
        case 3:
        case 4:
            writer.writeSignedBits(value, 16);
            break;

        case 5:
        case 6:
            break;

        default:
            writer.writeSignedBits(value, 32);
            break;
        }
    }

    static const uint16_t VARIANT_A_SELECTOR;
    static const uint16_t VARIANT_B_SELECTOR1;
    static const uint16_t VARIANT_B_SELECTOR2;
    static const uint16_t VARIANT_B_SELECTOR3;
    static const uint16_t EMPTY_SELECTOR1;
    static const uint16_t EMPTY_SELECTOR2;
    static const uint16_t VARIANT_C_SELECTOR;
};

const uint16_t IntChoiceTest::VARIANT_A_SELECTOR = 1;
const uint16_t IntChoiceTest::VARIANT_B_SELECTOR1 = 2;
const uint16_t IntChoiceTest::VARIANT_B_SELECTOR2 = 3;
const uint16_t IntChoiceTest::VARIANT_B_SELECTOR3 = 4;
const uint16_t IntChoiceTest::EMPTY_SELECTOR1 = 5;
const uint16_t IntChoiceTest::EMPTY_SELECTOR2 = 6;
const uint16_t IntChoiceTest::VARIANT_C_SELECTOR = 7;

TEST_F(IntChoiceTest, emptyConstructor)
{
    IntChoice intChoice;
    ASSERT_THROW(intChoice.getTag(), zserio::CppRuntimeException);
}

TEST_F(IntChoiceTest, bitStreamReaderConstructor)
{
    const uint16_t tag = VARIANT_A_SELECTOR;
    zserio::BitStreamWriter writer;
    const int8_t value = 99;
    writeIntChoiceToByteArray(writer, tag, value);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);

    const IntChoice intChoice(reader, tag);
    ASSERT_EQ(tag, intChoice.getTag());
    ASSERT_EQ(value, intChoice.getA());
}

TEST_F(IntChoiceTest, copyConstructor)
{
    const uint16_t tag = VARIANT_A_SELECTOR;
    IntChoice intChoice;
    intChoice.initialize(tag);

    const VariantA value = 99;
    intChoice.setA(value);

    const IntChoice intChoiceCopy(intChoice);
    ASSERT_EQ(tag, intChoiceCopy.getTag());
    ASSERT_EQ(value, intChoiceCopy.getA());
}

TEST_F(IntChoiceTest, operatorAssignment)
{
    const uint16_t tag = VARIANT_B_SELECTOR3;
    IntChoice intChoice;
    intChoice.initialize(tag);

    const VariantB value = 234;
    intChoice.setB(value);

    const IntChoice intChoiceCopy = intChoice;
    ASSERT_EQ(tag, intChoiceCopy.getTag());
    ASSERT_EQ(value, intChoiceCopy.getB());
}

TEST_F(IntChoiceTest, initialize)
{
    const uint16_t tag = EMPTY_SELECTOR1;
    IntChoice intChoice;
    intChoice.initialize(tag);
    ASSERT_EQ(tag, intChoice.getTag());
}

TEST_F(IntChoiceTest, getTag)
{
    const uint16_t tag = EMPTY_SELECTOR2;
    IntChoice intChoice;
    intChoice.initialize(tag);
    ASSERT_EQ(tag, intChoice.getTag());
}

TEST_F(IntChoiceTest, getSetA)
{
    IntChoice intChoice;
    intChoice.initialize(VARIANT_A_SELECTOR);

    const VariantA value = 99;
    intChoice.setA(value);
    ASSERT_EQ(value, intChoice.getA());
}

TEST_F(IntChoiceTest, getSetB)
{
    IntChoice intChoice;
    intChoice.initialize(VARIANT_B_SELECTOR2);

    const VariantB value = 234;
    intChoice.setB(value);
    ASSERT_EQ(value, intChoice.getB());
}

TEST_F(IntChoiceTest, getSetC)
{
    IntChoice intChoice;
    intChoice.initialize(VARIANT_C_SELECTOR);

    const VariantC value = 65535;
    intChoice.setC(value);
    ASSERT_EQ(value, intChoice.getC());
}

TEST_F(IntChoiceTest, bitSizeOf)
{
    IntChoice intChoice;
    intChoice.initialize(VARIANT_A_SELECTOR);
    const VariantA valueA = 99;
    intChoice.setA(valueA);
    ASSERT_EQ(8, intChoice.bitSizeOf());

    intChoice.initialize(VARIANT_B_SELECTOR2);
    const VariantB valueB = 234;
    intChoice.setB(valueB);
    ASSERT_EQ(16, intChoice.bitSizeOf());

    intChoice.initialize(EMPTY_SELECTOR2);
    ASSERT_EQ(0, intChoice.bitSizeOf());

    intChoice.initialize(VARIANT_C_SELECTOR);
    const VariantC valueC = 65535;
    intChoice.setC(valueC);
    ASSERT_EQ(32, intChoice.bitSizeOf());
}

TEST_F(IntChoiceTest, initializeOffsets)
{
    IntChoice intChoice;
    intChoice.initialize(VARIANT_A_SELECTOR);
    const size_t bitPosition = 1;
    ASSERT_EQ(9, intChoice.initializeOffsets(bitPosition));

    intChoice.initialize(VARIANT_B_SELECTOR1);
    ASSERT_EQ(17, intChoice.initializeOffsets(bitPosition));

    intChoice.initialize(EMPTY_SELECTOR1);
    ASSERT_EQ(1, intChoice.initializeOffsets(bitPosition));

    intChoice.initialize(VARIANT_C_SELECTOR);
    ASSERT_EQ(33, intChoice.initializeOffsets(bitPosition));
}

TEST_F(IntChoiceTest, operatorEquality)
{
    const uint16_t tag = VARIANT_A_SELECTOR;
    IntChoice intChoice1;
    intChoice1.initialize(tag);
    IntChoice intChoice2;
    intChoice2.initialize(tag);
    ASSERT_TRUE(intChoice1 == intChoice2);

    const VariantA valueA = 99;
    intChoice1.setA(valueA);
    ASSERT_FALSE(intChoice1 == intChoice2);

    intChoice2.setA(valueA);
    ASSERT_TRUE(intChoice1 == intChoice2);

    const VariantA diffValueA = valueA + 1;
    intChoice2.setA(diffValueA);
    ASSERT_FALSE(intChoice1 == intChoice2);
}

TEST_F(IntChoiceTest, hashCode)
{
    const uint16_t tag = VARIANT_A_SELECTOR;
    IntChoice intChoice1;
    intChoice1.initialize(tag);
    IntChoice intChoice2;
    intChoice2.initialize(tag);
    ASSERT_EQ(intChoice1.hashCode(), intChoice2.hashCode());

    const VariantA valueA = 99;
    intChoice1.setA(valueA);
    ASSERT_NE(intChoice1.hashCode(), intChoice2.hashCode());

    intChoice2.setA(valueA);
    ASSERT_EQ(intChoice1.hashCode(), intChoice2.hashCode());

    const VariantA diffValueA = valueA + 1;
    intChoice2.setA(diffValueA);
    ASSERT_NE(intChoice1.hashCode(), intChoice2.hashCode());
}

TEST_F(IntChoiceTest, read)
{
    const uint16_t tag = VARIANT_A_SELECTOR;
    IntChoice intChoice;
    intChoice.initialize(tag);

    zserio::BitStreamWriter writer;
    const int8_t value = 99;
    writeIntChoiceToByteArray(writer, tag, value);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    intChoice.read(reader);

    ASSERT_EQ(tag, intChoice.getTag());
    ASSERT_EQ(value, intChoice.getA());
}

TEST_F(IntChoiceTest, write)
{
    const uint16_t tagA = VARIANT_A_SELECTOR;
    IntChoice intChoice;
    intChoice.initialize(tagA);

    const VariantA valueA = 99;
    intChoice.setA(valueA);
    zserio::BitStreamWriter writerA;
    intChoice.write(writerA);
    size_t writeBufferByteSizeA;
    const uint8_t* writeBufferA = writerA.getWriteBuffer(writeBufferByteSizeA);
    zserio::BitStreamReader readerA(writeBufferA, writeBufferByteSizeA);
    IntChoice readIntChoiceA(readerA, tagA);
    ASSERT_EQ(valueA, readIntChoiceA.getA());

    const uint16_t tagB = VARIANT_B_SELECTOR2;
    intChoice.initialize(tagB);
    const VariantB valueB = 234;
    intChoice.setB(valueB);
    zserio::BitStreamWriter writerB;
    intChoice.write(writerB);
    size_t writeBufferByteSizeB;
    const uint8_t* writeBufferB = writerB.getWriteBuffer(writeBufferByteSizeB);
    zserio::BitStreamReader readerB(writeBufferB, writeBufferByteSizeB);
    IntChoice readIntChoiceB(readerB, tagB);
    ASSERT_EQ(valueB, readIntChoiceB.getB());

    const uint16_t tagC= VARIANT_C_SELECTOR;
    intChoice.initialize(tagC);
    const VariantC valueC = 65535;
    intChoice.setC(valueC);
    zserio::BitStreamWriter writerC;
    intChoice.write(writerC);
    size_t writeBufferByteSizeC;
    const uint8_t* writeBufferC = writerC.getWriteBuffer(writeBufferByteSizeC);
    zserio::BitStreamReader readerC(writeBufferC, writeBufferByteSizeC);
    IntChoice readIntChoiceC(readerC, tagC);
    ASSERT_EQ(valueC, readIntChoiceC.getC());
}

} // namespace int_choice
} // namespace choice_types
